multipass-addressing/src/main/scala/com/precisely/bigdata/sample/spark/MultipassAddressing.scala

```diff
@@ -0,0 +1,198 @@
+    addressInputDF
+      // Adds a new column, represented as a collection comprised of the outputFields and the error field
+      .withColumn("addressing_geocode_result", addressingUdf(map(
+        lit("addressLines[0]"), col("streetName"),
+        lit("city"), col("city"),
+        lit("admin1"), col("state"),
+        lit("postalCode"), col("zip")))
+      )
+      // Persist the addressing result to avoid recalculation when we expand the result
+      .persist()
+      // Expand the result collection such that each output field is a separate column, including the error field.
+      .select("*", "addressing_geocode_result.*").drop("addressing_geocode_result")
+      // Write the dataframe to the specified output folder as csv file
+      .write.mode(SaveMode.Overwrite).option("header", "true").csv(outputDirectory)
+  }
+}
+
+/*
+ * A custom Executor that contains the logic to perform verify and then multipass geocode
+ */
+class CustomExecutor extends AddressingExecutor {
+
+  import com.precisely.addressing.v1.model.Response
+  import com.precisely.addressing.v1.{Addressing, Preferences}
+
+  // set preference to return all fields and Match Mode to Relaxed
+  private val relaxedGeocoderPreferences: Preferences = new PreferencesBuilder()
+    .withReturnAllInfo(true)
+    .withMatchMode("RELAXED")
+    .build()
+
+  override def execute(input: RequestInput, preferences: Option[Preferences], addressing: Addressing): Response = {
+
+    val verifyResponse = addressing.verify(input.requestAddress(), preferences.orNull)
+
+    val inputRequest: RequestAddress = input.requestAddress()
+
+    var multiLineRequest: RequestAddress = inputRequest
+
+    if (CollectionUtils.isNotEmpty(verifyResponse.getResults)) {
+      val verifyAddress = verifyResponse.getResults.get(0).getAddress
+
+      //  create multiLine request from verify Response
+      multiLineRequest = new RequestAddress()
+      multiLineRequest.setAddressLines(getValidAddressLine(verifyResponse.getResults.get(0).getAddressLines))
+      multiLineRequest.setStreet(verifyAddress.getStreet)
+      multiLineRequest.setAddressNumber(verifyAddress.getAddressNumber)
+      multiLineRequest.setCity(verifyAddress.getCity.getLongName)
+      multiLineRequest.setPostalCode(verifyAddress.getPostalCode)
+      multiLineRequest.setCountry(verifyAddress.getCountry.getIsoAlpha3Code)
+
+    }
+
+    val multiLineResponse = addressing.geocode(multiLineRequest, relaxedGeocoderPreferences)
+    if (CollectionUtils.isNotEmpty(multiLineResponse.getResults) && isAddressLevelMatch(multiLineResponse.getResults.get(0))) {
+      return multiLineResponse
+    }
+
+    val singleLine: String = if (CollectionUtils.isNotEmpty(verifyResponse.getResults)) {
+      for (addressLine <- 0 to getValidAddressLine(verifyResponse.getResults.get(0).getAddressLines).size() - 1) {
+        verifyResponse.getResults.get(0).getAddressLines.get(addressLine) + " "
+      }
```

Comment:
Couple problems with these lines (130-132):
- each iteration of the for loop is performing a string concatenation, but doing nothing with it because you're doing no assignment.  The for loop is essentially doing nothing.
- getValidAddressLines will remove a number of lines from the results address lines, and the reason for that we talked about was because in some cases the first lines may be blank.  In this loop though, you're using the index based on the trimmed list and accessing elements of the untrimmed list.  So, if the inside of this list was actually calculating something, you'd potentially be looking at the wrong elements.  For example, if the original lines were {"", "", "line3", "line4","line5"} then your trimmed list used for iteration would look like: {"line3","line4"}.  And, since you're iterating over indexes in your trimmed list and pulling elements from your untrimmed list, then these indexes (0 and 1) would both be blank in your calculation.  A better approach here if you wanted to combine the elements returned from `getValidAddressLine` would have been something like: `getValidAddressLine(verifyResponse.getResults.get(0)).mkString(" ")`
- like the above bullet, about the indexes, another line removed by getAddressLines is the last line of the address which is needed to create the single line address.  So, if we use the "valid" method (which i mention below we should rename anyway to "getMainAddressLines") then we'er going to be missing out on important info